service/controller/appcatalog/resource/appcatalogentry/create.go

```diff
@@ -168,154 +170,178 @@ func (r *Resource) updateAppCatalogEntry(ctx context.Context, entryCR *v1alpha1.
+		v, err := semver.NewVersion(entries[i].Version)
+		if errors.As(err, &semver.ErrInvalidSemVer) {
+			r.logger.Debugf(ctx, "invalid semver from converting app entries %s, version is %s", entries[i].Name, entries[i].Version)
+			continue
+		} else if err != nil {
+			return entry{}, microerror.Mask(err)
+		}
 
-			if entryCR.Spec.Chart.APIVersion == "" {
-				// chartAPIVersion default is `v1`.
-				entryCR.Spec.Chart.APIVersion = "v1"
-			}
+		// Removing Prerelease from version since they are mostly SHA strings which we could not compare the size.
+		nextVersion, err := v.SetPrerelease("")
+		if err != nil {
+			return entry{}, microerror.Mask(err)
+		}
 
-			if entryCR.Spec.DateCreated == nil {
-				// If meta.yaml does not have dateCreated, use the timestamp from app.
-				entryCR.Spec.DateCreated = &e.Created
-				entryCR.Spec.DateUpdated = updatedTime
-			}
+		if nextVersion.GreaterThan(&latestVersion) {
+			latestIndex = i
+			latestVersion = nextVersion
+			latestCreated = entries[i].Created.DeepCopy()
+			continue
+		}
 
-			entryCRs[name] = entryCR
+		if nextVersion.Equal(&latestVersion) {
+			if latestCreated != nil && latestCreated.After(entries[i].Created.Time) {
+				latestIndex = i
+				latestVersion = nextVersion
+				latestCreated = entries[i].Created.DeepCopy()
+			}
 		}
 	}
 
-	return entryCRs, nil
+	return entries[latestIndex], nil
 }
 
-func sortEntry(entries []entry) ([]entry, error) {
-	if len(entries) == 0 {
-		return nil, nil
-	}
+func (r *Resource) newAppCatalogEntries(ctx context.Context, cr v1alpha1.AppCatalog, index index) (map[string]*v1alpha1.AppCatalogEntry, error) {
+	entryCRs := map[string]*v1alpha1.AppCatalogEntry{}
 
-	sort.Slice(entries, func(i, j int) bool {
-		// Removing Prerelease from version since they are mostly SHA strings which we could not compare the size.
-		prevRelease, err := entries[i].Version.SetPrerelease("")
-		if err != nil {
-			return false
+	for _, entries := range index.Entries {
+		sort.Slice(entries, func(i, j int) bool {
+			return entries[i].Created.After(entries[j].Created.Time)
+		})
+
+		maxEntries := r.maxEntriesPerApp
+		if len(entries) < maxEntries {
+			maxEntries = len(entries)
 		}
-		nextRelease, err := entries[j].Version.SetPrerelease("")
+
+		latestEntry, err := r.getLatestEntry(ctx, entries)
 		if err != nil {
-			return false
+			return nil, microerror.Mask(err)
 		}
 
-		if prevRelease.GreaterThan(&nextRelease) {
-			return true
-		}
+		for i := 0; i < maxEntries; i++ {
+			e := entries[i]
 
-		if prevRelease.Equal(&nextRelease) {
-			if entries[i].Created.After(entries[j].Created.Time) {
-				return true
+			entryCR, err := r.getDesiredAppCatalogEntryCR(ctx, &cr, e, latestEntry.Version == e.Version)
+			if err != nil {
+				return nil, microerror.Mask(err)
 			}
+
+			entryCRs[entryCR.Name] = entryCR
 		}
 
-		return false
-	})
+		// If the latest entry is not included in the desired CRs, we should add them so users could see the latest CR always.
```

Comment:
```suggestion
		// If the latest entry is not included in the desired CRs, we add it so users can always see the latest CR.
```

Thanks for adding <3